### PR TITLE
New deployment system

### PR DESCRIPTION
**Pull request by siddharthist**
This enables the deployment and testing of MI on all supported platforms with minimal user intervention. It will provide a base for Jenkins-based automated testing of the platform. 

Extensive documentation is available in the README. The code is also commented thoroughly, and should be easily readable. 

There is clearly more work to do, with a lot of TODOs in the code and README, but it works right now and should be quite helpful in the meantime.

### CODE REVIEW

**Comment by siddharthist** [comment]
@stevendborrelli Was discussing spinning this off into an Ansible/Terraform testing suite, possibly integrated with Distributive's network testing capabilities and terraform.py's .tfstate parsing power. Further discussion is needed.

**Comment by siddharthist** [comment]
This will eventually be part of a separate project, but for now we can use it to deploy more rapidly.

**Comment by siddharthist** [comment]
Related: #163, #547

**Comment by saswatp** [comment]
@siddharthist - Will mi-deploy support multi-cloud deployment with different cloud platforms ? For example , a set of control nodes in Openstack and another set in AWS with Consul WAN discovery enabled. Would like to know if this use case has been tested.

**Comment by siddharthist** [comment]
@saswatp That hasn't been tested, but I'd love feedback. Again, this isn't mature or permanent, just a stop-gap for testing deployments. It is not how MI should be deployed for production.

**Comment by saswatp** [comment]
Thanks @siddharthist . mi-deploy is one feature that I was really looking forward to. We were planning to something similar and now that it exists we would be happy to contribute to. 

Multi cloud deployment -  You could be using an IaaS that doesn't have presence in all regions and your application service needs that presence. Your primary data center and new data center could be built on different IaaS .  That is when this feature will be useful . I will do a test, but not sure when I could start on that. Once I have something , I will let you know.

**Comment by siddharthist** [comment]
@saswatp Sounds awesome. Once we get a more general solution (Ansible + Terraform deployment and testing), I'll send a link to that work your way.

**Comment by ohal** [comment]
it will be perfect if somebody take a look into it https://github.com/CiscoCloud/microservices-infrastructure/issues/163#issuecomment-111712098

mentioned automated testing, or kind of, is already done and here documented https://github.com/CiscoCloud/microservices-infrastructure/blob/qa/features-master-integration/docs/jenkins_ci.md
obviously, it's not full framework testing, however we're able to move forward

**Comment by siddharthist** [comment]
@ohal There are a couple of reasons we are starting to move to Distributive-based testing
 1) That Jenkins build fails all the time on commits we test and are stable
 2) Distributive also provides a layer of runtime testing for the end user, which Jenkins alone cannot. It provides additional value to just CI testing
 3) We want to develop a more general tool for testing cloud deployments on a variety of providers (AWS, GCE, DO, OS, Vagrant), instead of limiting testing to one cloud of having a constantly running Jenkins instance on all of our supported platforms.

**Comment by eirslett** [comment]
Hey! We're testing this on softlayer, looks like it works fine there as well.
But we're running a custom-built version of mi-deploy, because it doesn't allow "--platform softlayer". So we added that to the list of valid --platform arguments in our fork of mi-deploy. Could you please do the same in this PR, so it works for us too? :grinning:

**Comment by siddharthist** [comment]
@eirslett Absolutely! That's great to hear. To make sure I've got this right, it works the same as with AWS, OS, etc., where you just have a softlayer.tf file that gets copied in? Anything different that I should keep in mind? Would you like to post your custom code in a Gist, so I can just add it to this PR?

**Comment by eirslett** [comment]
Exactly! Just search for "digitalocean" in this PR; where that is added as an option, add "softlayer" as well. Then it will use "softlayer.tf".

**Comment by eirslett** [comment]
This is our diff: https://github.com/CiscoCloud/microservices-infrastructure/pull/647

File: mi-deploy/mi-deploy.go [hunk 1]
```diff
@@ -0,0 +1,553 @@
...
+// but doesn't lend itself to breaking down.
+// TODO app action doesn't seem to be run every time and set the verbosity.
+// CLI might be able to use a PR.
+func getApp() (app *cli.App) {
+	// global app variables (name, version, etc)
+	app = cli.NewApp()
+	app.Name = "mi-deploy"
+	app.Usage = "Deploy MI to various public clouds"
+	app.Version = "0.0.1-dev"
+	app.Author = "Langston Barrett"
```
Comment (stevendborrelli) [inline mi-deploy/mi-deploy.go]: Add your email address.

**Comment by siddharthist** [comment]
@stevendborrelli Done!

**Comment by stevendborrelli** [comment]
Some feedback:

1. Even in debug verbosity, output isn't displayed from ansible & terraform
2. Make inline help easier. Right now the `--help` option doesn't show the options for running the deployment. 
3. Maybe run checks for ansible & directories existing before you actually git clone or terraform plan.

**Comment by eirslett** [comment]
It needs a --non-interactive option, so it can be run from a CI system/jenkins and return a non-zero exit code it the deploy fails.

**Comment by stevendborrelli** [comment]
It would be nice to use a .yml file to orchestrate the different deployments, something like travis ci: http://docs.travis-ci.com/user/customizing-the-build/

**Comment by siddharthist** [comment]
@stevendborrelli @eirslett I definitely thought a lot about both of those, and see them as absolutely necessary future features. I'd love to continue work on this as a separate Ansible/Terraform deployment+testing system. @stevendborrelli Would you like me to close this PR until I have those features up and running, and put it in a different repo? Include them here? What do you think would be best way forward?

**Comment by stevendborrelli** [comment]
I think the main feature I'd like to see if the output when running. mi-deploy can be merged here, but I think as it becomes more general we can move it it's own repository.

**Comment by saswatp** [comment]
Was trying to build mi-deploy and ran into an issue. 

mi-deploy]> go get . && go build .
go install: no install location for directory /Users/saswat/mi_deploy/microservices-infrastructure/mi-deploy outside GOPATH

/Users/saspraha/mi_deploy/microservices-infrastructure/mi-deploy/packages

Should the user have to change his GOPATH for this to work ?

**Comment by PeterLamar** [comment]
Maybe I am jumping the gun, but I think this is a great idea and want to start discussing what this would look like in its own repo. Here is a strawman to improve on? 

https://github.com/CiscoCloud/deploy

**Comment by siddharthist** [comment]
@PeterLamar We've begun discussions on a Mantl CLI, I'm happy to hear that you guys are excited :)

**Comment by ohal** [comment]
test this please

**Comment by PeterLamar** [comment]
@siddharthist Great! Lets see some designs :) Feel free to take over the repo I created, rename or start something new

**Comment by eirslett** [comment]
I had a look at Mantl now.
We'll run ~8 different in-house mesos clusters (different environments, and multi-dc), and what works quite well for us (so far), is to have a "master" config repo for all the clusters, with a subdirectory for every cluster. So you `cd cluster-name && mi-deploy deploy [arguments]`.
I think configuration files are better than command line arguments; they can be checked into version control. Chances are you want to deploy cluster X on the same cloud provider that you did last time (for example).

We have a use case where we deploy the same setup to two different cloud providers (which is for integration testing; we test the development branch against both SoftLayer and OpenStack), but then we just duplicate the directories and configations. `cd softlayer-integration-test && mi-deploy deploy, cd openstack-integration-test && mi-deploy deploy`

What I think would be the easiest/most minimal setup for mi-deploy (or Mantl) is to simply deploy a directory. `mi-deploy deploy .` Or a subdirectory `mi-deploy deploy my-cluster`, which deploys whatever is inside the "my-cluster" subdirectory. (Which is a couple of configuration files)

That's one way mi-deploy could be changed; instead of having --provider=xyz, it just copies whatever terraform files it finds in the `tf-files` folder and that's it.

**Comment by cdmitri** [comment]
All - well be working with Hortonworks to get Ambari to understand Mesos
and Mantl in particular. It already has lots of things that we need
including blueprints (config) cicd, some monitoring, etc. Check it
out...let's see where the gaps are :)

**Comment by siddharthist** [comment]
@stevendborrelli I just added continuous output on debug verbosity, check it out and let me know what you think!

**Comment by siddharthist** [comment]
@stevendborrelli Is there anything else you'd like me to change before we do an initial merge and start work on a real CLI?

**Comment by eirslett** [comment]
@siddharthist if you find me on Gitter I can give you some feedback on the deployment system - we're using it (a modified version of it) to deploy Mantl to test and production environments.

**Pull request merged**